### ISSUE TITLE
rate_limit on Cisco, Arista, Sonic, Dell OS 10

### ISSUE THREAD

**Issue by juztas** [sdn-sense/siterm#440]
Dell OS 9 and SiteRM publishes information right now for qos (https://github.com/sdn-sense/siterm/issues/189) - Need to look on how to do this on different devices which SiteRM Supports at given moment. 

- [x] Arista
- [ ] Sonic
- [ ] Dell OS 10
- [ ] Cisco
- [ ] Juniper

**Comment by juztas** [comment]
Dell OS 10
```

DELL OS10 (> 10.5.3 Release)
class-map match-any VLAN3618
  # police cir 4000000 bc 200000 pir 400000
  match vlan 3618
class-map match-any VLAN3615
  match vlan 3615
class-map match-any VLAN3619
  match vlan 3619

policy-map type qos SENSE_QOS
  class VLAN3615
    set qos-group 1
  class VLAN3618
    set qos-group 4
  class VLAN3619
    set qos-group 7
 interface ethernet 1/1/21
service-policy input type qos SENSE_QOS
 interface ethernet 1/1/22
service-policy input type qos SENSE_QOS
 interface ethernet 1/1/23
service-policy input type qos SENSE_QOS
 interface ethernet 1/1/24
service-policy input type qos SENSE_QOS
```

**Comment by juztas** [comment]
Dell OS 9 is limited in its capabilities and mainly expects DSCP (which would require rewriting on the client side). It will not be supported moving forward.

**Comment by juztas** [comment]
1 - bestEffort;
4 - softCapped;
7 - guaranteedCapped;
Allow this to be configured per site configuration. (Is this something we should expose to SENSE?)

**Comment by juztas** [comment]
LA Arista changes done here:
https://github.com/sdn-sense/sense-aristaeos-collection/commit/cc69c0b1b14c64baff12efd93409f499969d007c - mainly to send commands in batch (to speed up apply, this is a minimal improvement, but helps if the switch/router is far away)
https://github.com/sdn-sense/ansible-templates/commit/02148fc2114ee16e95a0ae29cf825fdca66db456 - Jinja template update for QoS on Arista;
------
SiteRM changes (general for all devices): https://github.com/sdn-sense/siterm/compare/qoscalculateremain?expand=1
SiteRM changes include: Move traffic_classes inside qos_policy (under switch). Allow to specify max_policy_rate and burst_size.

**Comment by juztas** [comment]
Here is how QoS works on SiteRM (on Arista).  
The default configuration specifies the following parameters, which can be overridden by the RM config:  

```json
"qos_policy": {
    "traffic_classes": {
        "default": 1,
        "bestEffort": 2,
        "softCapped": 4,
        "guaranteedCapped": 7
    },
    "max_policy_rate": "268000",
    "burst_size": "256"
}
```

Arista has certain limits. For example:  
- **max_policy_rate** is the maximum `police rate` that can be set for a class.  
- **burst_size** is the maximum burst size.  

These parameters are not provided in the request. (In the future, it would be beneficial if the Orchestrator calculated the burst size based on the RTT and MTU.)  

SENSE can request one of the following bandwidth types: **guaranteedCapped**, **softCapped**, or **bestEffort**.  

When SiteRM receives a request, it calculates the bandwidth for all requests based on the following parameters:  

1. **Available port capacity** (can be overridden by RM config if the site wants to reserve a certain percentage of bandwidth for other purposes).  

2. If the request is **guaranteedCapped**, SiteRM will add the following configuration on the Arista device:  
   ```
   policy-map type quality-of-service SENSE_QOS
      class VLAN{VLAN_ID}
         set traffic-class {TRAFFICCLASSGC}
         police rate {GC_REQUEST_BW} mbps burst-size {BURST_SIZE} mbytes
   ```

3. After all **guaranteedCapped** allocations are configured, SiteRM calculates the remaining bandwidth:  
   ```
   {NewRemainingCapacity} = {AvailablePortCapacity} - {ActiveGuaranteedCapped}
   ```

4. For **softCapped**, SiteRM will configure:  
   ```
      class VLAN{VLAN_ID}
         set traffic-class {TRAFFICCLASSSC}
         police rate {SC_REQUEST_BW} mbps burst-size {BURST_SIZE} mbytes \
            action set drop-precedence rate {NewRemainingCapacity} mbps burst-size {BURST_SIZE} mbytes
   ```

5. For **bestEffort**, SiteRM will configure (always allowing at least 100 Mbps, up to the maximum remaining capacity):  
   ```
      class VLAN{VLAN_ID}
         set traffic-class {TRAFFICCLASSBE}
         police rate 100 mbps burst-size {BURST_SIZE} mbytes \
            action set drop-precedence rate {NewRemainingCapacity} mbps burst-size {BURST_SIZE} mbytes
   !
   ```

**Note:** SoftCapped and BestEffort are very similar in that both can reach the maximum remaining capacity of the link. However, if both compete for bandwidth simultaneously, traffic class prioritization applies (2 vs 4), and BestEffort has lower priority than SoftCapped.

**Comment by xi-yang** [comment]
Three questions:
1. Does the committed packets of softCapped have lower priority than guaranteedCapped (4 vs 7)? So there is a chance the committed packets of softCapped will be dropped when competing with the guaranteedCapped if the accounting of total bandwidth isn't done correctly. 
2. The maximum commit rate is 288G per flow, correct?
3. What is the point of having 100mbps committed for bestEffort if it has the lowest priority that would get dropped anyways when competing with others?

**Comment by juztas** [comment]
1. Correct, guaranteedCapped has a higher priority than softCapped/bestEffort. SoftCapped will not be dropped, as it will not compete with guaranteedCapped in the current implementation. SiteRM knows the total available capacity (let's say 100Gbps), and if the request is 20Gbps guaranteed, the remaining BW for softQoS is 80Gbps max. Unless SiteRM fails to compute it correctly (and fails to apply the correct policy rate), then it is up to the arista to prioritize based on traffic-class (guaranteed - 7, softCapped -4, bestEffort 1). If that happens, then Guaranteed Capped could easily take it over everything. But that is where SENSE shines, that we also have traffic shaping at the host level, and it tries to be a "good citizen", not pushing too much.
2. No, on Arista EOS, max_policy_rate is not per flow; it is applied per traffic class and match criteria in the policy map (in our case, per vlan)
3. Just my thinking of implementation. Even though it is a bestEffort, it does allocate a minimum of 100 Mbps. This might change as we start some initial testing.

**Comment by juztas** [comment]
Correct, any request above 268 (Arista limit), will set a hard limit to 268gbps. In this case, the best approach is to use multipath BGP (split to be below 268 for each VLAN)

**Comment by juztas** [comment]
Host DSCP:
This configuration below uses Linux tc (traffic control) with u32 classifiers and pedit actions to rewrite the DSCP field for all traffic on vlan.1409.
Three classes are enforced:

- GuaranteedCapped → DSCP 46 (0x2E, encoded as 0xB8 in IPv4 TOS)
- SoftCapped → DSCP 18 (0x12, encoded as 0x48 in IPv4 TOS)
- BestEffort → DSCP 0

Because IPv4 and IPv6 store DSCP in different spots, each class has two filters: one for IPv4 and one for IPv6.

```
GuaranteedCapped: (DSCP 46)
tc filter add dev vlan.1409 parent 1: protocol ipv6 prio 11 u32 match u32 0 0 action pedit munge offset 0 u8 set 0x6e action pedit munge offset 1 u8 set 0x80 action ok
tc filter add dev vlan.1409 parent 1: protocol ip prio 1 u32 match u32 0 0 action pedit munge ip tos set 0xb8 action ok
SoftCapped: (DSCP 18)
tc filter add dev vlan.1409 parent 1: protocol ip prio 2 u32 match u32 0 0 action pedit munge ip tos set 0x48 action ok
tc filter add dev vlan.1409 parent 1: protocol ipv6 prio 12 u32 match u32 0 0 action pedit munge offset 0 u8 set 0x48 action pedit munge offset 1 u8 set 0x00 action ok
BestEffort: (DSCP 0)
tc filter add dev vlan.1409 parent 1: protocol ip prio 3 u32 match u32 0 0 action pedit munge ip tos set 0x00 action ok
tc filter add dev vlan.1409 parent 1: protocol ipv6 prio 13 u32 match u32 0 0 action pedit munge offset 0 u8 set 0x00 action pedit munge offset 1 u8 set 0x00 action ok
```
This will require to match QoS class and apply required DSCP at the host. This covers only L2 (vlan connected directly to the host). Need to continue to research for rules for L3 DSCP set. This is only for host to set it, routers/switches to respect QoS is another research topic.
